### PR TITLE
HPCC-18622 Store dictionary payload information in RtlTypeInfo structures

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Regression suites, compiler suite.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday This is the work I did to get payload info out on dictionary structures - totally independent of the work on indexes, and may well be barking up entirely the wrong tree.

File: ecl/hqlcpp/hqlhtcpp.cpp [hunk 1]
```diff
@@ -3584,39 +3584,50 @@ IHqlExpression * HqlCppTranslator::getRtlFieldKey(IHqlExpression * expr, IHqlExp
...
+        AColumnInfo * root = map->queryRootColumn();
+        AColumnInfo * columnInfo = root->lookupColumn(expr);
+
+        if (columnInfo->isPayloadField())
         {
-        case type_bitfield:
+            isPayload = true;
+            contextDependent = true;
+        }
+        else
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlhtcpp.cpp]: not sure this should be else.  bitfields in payloads still need the code that follows

File: ecl/hqlcpp/hqltcppc.cpp [hunk 2]
```diff
@@ -3358,16 +3373,32 @@ bool ColumnToOffsetMap::isMaxSizeSpecified()
...
 {
     assertex(record->getOperator() == no_record);
     unsigned max = record->numChildren();
     unsigned idx;
+    unsigned firstPayloadField = max;
     bool fixedSize = true;
     prior = NULL;
+    if (payloadFields)
+    {
+        unsigned payloadSeen = 0;
```
Comment (ghalliday) [inline ecl/hqlcpp/hqltcppc.cpp]: Should probably be in a shared function - could be called from elsewhere

File: ecl/hqlcpp/hqltcppc.cpp [hunk 3]
```diff
@@ -3740,7 +3771,6 @@ IHqlExpression * SerializationRow::createField(IIdAtom * id, ITypeInfo * type)
         return createField(NULL, type);                     // name clash -> create a new unnamed field
 
     queryRecord()->addOperand(LINK(newField));
-    serializedMap->addColumn(newField, translator.queryRecordMap());
```
Comment (ghalliday) [inline ecl/hqlcpp/hqltcppc.cpp]: was this deleted on purpose?

**Comment by ghalliday** [comment]
I think the change is reasonable.  I'm not a great fan of the context dependent flag - better would be to add a payload attribute onto the fields, but I suspect that will cause a more problems, with little benefit

**Comment by ghalliday** [comment]
checkboxes please

**Comment by richardkchapman** [comment]
@ghalliday Checkboxes done

**Pull request merged**